# Assignment #4 - Data Gathering and Warehousing - DSSA-5102

Instructor: Melissa Laurino</br>
Spring 2025</br>

Name: Thompson Morgan
</br>
Date: 2/20/25
<br>
<br>
<b>Only Murders in the...Database?</b><br>
An introduction to navigating databases using SQL, Python and Jupyter Notebook. <br>
<br>
A crime has taken place and you are the perfect amateur detective to solve it! 
<br>
The detective gave you the crime scene report, but you somehow lost it...? You vaguely remember that the crime
was a murder that occurred sometime on <b>Jan.15, 2018</b> and that it took place in <b>SQL
City</b>. All the clues to this mystery are buried in a huge database, and you need to use
SQL to navigate through this network of information. Your first step to solving the mystery is to retrieve the corresponding crime scene report from the police department’s database. Take a look at the cheatsheet to learn how to do this! From there, you can use your SQL skills to find the murderer.<br>

Your task for <b>Assignment #4</b> is to:<br>
- Connect to the database stored locally using Python.
- Explore the database by listing all tables, fields within the tables, and data types.
- Find out who the murderer is!<br>
- Add detailed comments to explain EVERY query or SQL command you use while we are still learning and practicing. I have my steps outlined, but please add more cells inbetween for additional queries! There is no limit on the number of queries you can use. <br>
- For each query include comments such as "SELECT all records FROM table WHERE column name = X"<br>
- Answer any prompts in markdown cells.<br>
<br>

This fabulous database was created by @NUKnightLab on Github and can be found here: https://github.com/NUKnightLab/sql-mysteries

Database files are indicated by the .db file extension. For this exercise, our database has already been created. To avoid setting your working directory, have the .db file and Assignment#4_LastName.ipynb in the same file location. There are multiple libraries that can be used to navigate database files and connect to SQL for databases that are stored on a server or your computer as a local host.

In [4]:
# Load necessary packages:
from sqlalchemy import create_engine, inspect, text # Database navigation
import sqlite3 # A second option for working with databases
import pandas as pd # Python data manilpulation

In [5]:
# How to connect to our .db file using library SQLAlchemy
# Notice that we are not specifiying a host, password, or server since this .db file is stored locally.
db_path = "sql-murder-mystery.db"  # Change this to your actual file name
engine = create_engine(f"sqlite:///{db_path}") # Use the create_engine function to connect to the database
engine # The f is the connection string, so if we print it we can see below

Engine(sqlite:///sql-murder-mystery.db)

SQLAlchemy needs to know what kind of database we are accessing. .db files can be treated as a SQLite database. Other options include databases formatted for MySQL, PostgreSQL, etc. Database type prefixes must be included.

In [7]:
# Explore the database

# Initializing the inspector command
inspector = inspect(engine)

# List the tables in our database:
for table_name in inspector.get_table_names():
    print("Table: " + table_name)

# So much data! How would we possibly make sense of ANY of this without SQL? :)

Table: crime_scene_report
Table: drivers_license
Table: facebook_event_checkin
Table: get_fit_now_check_in
Table: get_fit_now_member
Table: income
Table: interview
Table: person
Table: solution


In [8]:
# What are the column names for each table? What table will help us get to the next clue?
for table_name in inspector.get_table_names():
   for column in inspector.get_columns(table_name):
       print("Table: " + table_name + ", Column: " + column['name'] + ", Type: ", column['type'])

Table: crime_scene_report, Column: date, Type:  INTEGER
Table: crime_scene_report, Column: type, Type:  TEXT
Table: crime_scene_report, Column: description, Type:  TEXT
Table: crime_scene_report, Column: city, Type:  TEXT
Table: drivers_license, Column: id, Type:  INTEGER
Table: drivers_license, Column: age, Type:  INTEGER
Table: drivers_license, Column: height, Type:  INTEGER
Table: drivers_license, Column: eye_color, Type:  TEXT
Table: drivers_license, Column: hair_color, Type:  TEXT
Table: drivers_license, Column: gender, Type:  TEXT
Table: drivers_license, Column: plate_number, Type:  TEXT
Table: drivers_license, Column: car_make, Type:  TEXT
Table: drivers_license, Column: car_model, Type:  TEXT
Table: facebook_event_checkin, Column: person_id, Type:  INTEGER
Table: facebook_event_checkin, Column: event_id, Type:  INTEGER
Table: facebook_event_checkin, Column: event_name, Type:  TEXT
Table: facebook_event_checkin, Column: date, Type:  INTEGER
Table: get_fit_now_check_in, Column: m

Practice using commands useful in queries:

In [10]:
# Example: 
# Select all records from the table crime_scene_report where type is murder:

with engine.connect() as connection:  # Establish a connection
    query = text("""SELECT * 
                    FROM crime_scene_report 
                    WHERE type='murder'
                    """)                      # Define the query - text() ensures that the query string is read as a SQL expression 
    practice = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database
    
# Print the results
practice

date    type                                        description  \
0    20180115  murder                  Life? Dont talk to me about life.   
1    20180115  murder  Mama, I killed a man, put a gun against his he...   
2    20180215  murder                         REDACTED REDACTED REDACTED   
3    20180215  murder  Someone killed the guard! He took an arrow to ...   
4    20170107  murder  ‘It proves nothing of the sort!’ said Alice. ‘...   
..        ...     ...                                                ...   
143  20180427  murder              ‘Only a thimble,’ said Alice sadly.\n   
144  20180428  murder  wag my tail when I’m angry. Therefore I’m mad.’\n   
145  20180429  murder  (she was obliged to say ‘creatures,’ you see, ...   
146  20180430  murder               ‘But what am I to do?’ said Alice.\n   
147  20180115  murder  Security footage shows that there were 2 witne...   

               city  
0            Albany  
1              Reno  
2          SQL City  
3          SQL City  
4           Memphis  
..              ...  
143  St. Petersburg  
144        Appleton  
145          Toledo  
146         Spokane  
147        SQL City  

[148 rows x 4 columns]

In [11]:
# Select all records from the table drivers_license where the eye color is blue:

with engine.connect() as connection:   # Establish a connection
    query = text("""SELECT * 
                    FROM drivers_license 
                    WHERE eye_color='blue'
                    """) # Define the query
    practice2 = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database

# Print the results
practice2

id  age  height eye_color hair_color  gender plate_number  \
0     101255   18      79      blue       grey  female       5162Z1   
1     101494   48      55      blue        red  female       81X1N7   
2     101611   40      65      blue      white  female       5O8VW7   
3     101726   78      81      blue       blue  female       236T17   
4     101836   56      51      blue       grey  female       3VSG76   
...      ...  ...     ...       ...        ...     ...          ...   
2020  996697   68      73      blue      black  female       QVEW13   
2021  997724   65      73      blue     blonde    male       640570   
2022  997950   84      69      blue       grey  female       66DP8D   
2023  998114   46      50      blue      brown  female       I0GI33   
2024  998681   65      67      blue      green  female       7N21H3   

           car_make       car_model  
0             Lexus              GS  
1               Kia        Sportage  
2               GMC   Sierra Denali  
3        Oldsmobile        Intrigue  
4           Hyundai           Azera  
...             ...             ...  
2020        Ferrari  612 Scaglietti  
2021         Toyota      Highlander  
2022       Infiniti             G37  
2023  Mercedes-Benz       SLK-Class  
2024       Chrysler           Aspen  

[2025 rows x 9 columns]

In [12]:
# Select distinct records of eye_color from the table drivers_license:

with engine.connect() as connection:   # Establish a connection
    query = text("""SELECT DISTINCT eye_color
                    FROM drivers_license
                    """) # Define the query
    practice3 = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database

# Print the results
practice3

eye_color
0     brown
1     green
2     amber
3      blue
4     black

In [13]:
# Select distinct records of name from the table person:

with engine.connect() as connection:   # Establish a connection
    query = text("""SELECT DISTINCT name
                    FROM person
                    """) # Define the query
    practice4 = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database

# Print the results
practice4

name
0       Christoper Peteuil
1      Kourtney Calderwood
2                Muoi Cary
3              Era Moselle
4             Trena Hornby
...                    ...
10006          Luba Benser
10007      Roxana Mckimley
10008      Cherie Zeimantz
10009          Allen Cruse
10010         Vance Hunten

[10011 rows x 1 columns]

In [14]:
# Select distinct records of gender AND car_model from the table drivers_license:

with engine.connect() as connection:   # Establish a connection
    query = text("""SELECT DISTINCT gender, car_model
                    FROM drivers_license
                    """) # Define the query
    practice5 = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database

# Print the results
practice5

gender car_model
0       male       MDX
1     female       SRX
2     female        xB
3     female     Rogue
4     female        GS
...      ...       ...
1365  female       550
1366  female   Prius c
1367  female       BRZ
1368    male   Mark LT
1369  female     HED-5

[1370 rows x 2 columns]

In [15]:
# Select distinct records of type and city from the table crime_scene_report:

with engine.connect() as connection:   # Establish a connection
    query = text("""SELECT DISTINCT type, city
                    FROM crime_scene_report
                    """) # Define the query
    practice6 = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database

# Print the results
practice6

type          city
0     robbery           NYC
1      murder        Albany
2      murder          Reno
3      murder      SQL City
4       theft       Chicago
...       ...           ...
1038  robbery       Trenton
1039  bribery  Garden Grove
1040    fraud         Houma
1041  assault       Fontana
1042  assault       Trenton

[1043 rows x 2 columns]

In [16]:
# Select all records from table drivers_license where the age is == 40:

with engine.connect() as connection:   # Establish a connection
    query = text("""SELECT *
                    FROM drivers_license
                    WHERE age == 40
                    """) # Define the query
    practice7 = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database

# Print the results
practice7

id  age  height eye_color hair_color  gender plate_number  \
0    101611   40      65      blue      white  female       5O8VW7   
1    103725   40      57     black      green  female       81WZ22   
2    116863   40      63     black      black  female       1ZSI8C   
3    118332   40      75     green      black    male       6SG3W0   
4    122134   40      61     black      white  female       8616Y4   
..      ...  ...     ...       ...        ...     ...          ...   
143  969815   40      63     green      black  female       G40MZY   
144  973824   40      50     brown       blue  female       LEJ771   
145  974364   40      52     green      black  female       822S0D   
146  990554   40      63     amber      white    male       XH2QC0   
147  992954   40      53     black     blonde  female       1QX3S8   

       car_make       car_model  
0           GMC   Sierra Denali  
1    Land Rover      Freelander  
2          Audi              A8  
3    Mitsubishi        Diamante  
4     Chevrolet           Astro  
..          ...             ...  
143   Chevrolet   Suburban 2500  
144        Ford            F250  
145        Ford            E250  
146    Chrysler  Town & Country  
147       Lexus              IS  

[148 rows x 9 columns]

In [17]:
# Select all records from table drivers_license where the age is greater than 21:

with engine.connect() as connection:   # Establish a connection
    query = text("""SELECT *
                    FROM drivers_license
                    WHERE age > 21
                    """) # Define the query
    practice8 = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database

# Print the results
practice8

id  age  height eye_color hair_color  gender plate_number  \
0     100280   72      57     brown        red    male       P24L4U   
1     100460   63      72     brown      brown  female       XF02T6   
2     101029   62      74     green      green  female       VKY5KR   
3     101198   43      54     amber      brown  female       Y5NZ08   
4     101494   48      55      blue        red  female       81X1N7   
...      ...  ...     ...       ...        ...     ...          ...   
9434  999509   42      80     amber       blue  female       8E403S   
9435  999536   56      65     green      green  female       4FFUD5   
9436  999940   71      61     green      green    male       1B8QN8   
9437  999981   67      69     brown       blue  female       1684K3   
9438  999986   49      58     green       grey    male       F8F64H   

        car_make car_model  
0          Acura       MDX  
1       Cadillac       SRX  
2          Scion        xB  
3         Nissan     Rogue  
4            Kia  Sportage  
...          ...       ...  
9434      Morgan    Aero 8  
9435     Mercury     Sable  
9436  Mitsubishi   Eclipse  
9437  Land Rover       LR2  
9438       Lexus        LS  

[9439 rows x 9 columns]

In [18]:
# Select all records from table drivers_license where the age is less than OR equal to 21:
# Do you see a difference in the number of rows?

with engine.connect() as connection:   # Establish a connection
    query = text("""SELECT *
                    FROM drivers_license
                    WHERE age <= 21
                    """) # Define the query
    practice8 = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database

# Print the results
practice8

id  age  height eye_color hair_color  gender plate_number  \
0    101255   18      79      blue       grey  female       5162Z1   
1    102536   21      76      blue     blonde    male       3DG322   
2    105064   20      53     black       blue    male       25UO1B   
3    108374   18      63     brown        red    male       X2KE6N   
4    109617   19      76     green       blue    male       UO2B77   
..      ...  ...     ...       ...        ...     ...          ...   
563  992923   20      77     black      brown  female       GT0EX0   
564  996996   20      56     black      green  female       5Y0W6U   
565  998350   20      78     amber       grey    male       6HRAXP   
566  999923   19      77     amber      black  female       5L0ZI4   
567  999993   18      63     black      black  female       6UZO2O   

          car_make     car_model  
0            Lexus            GS  
1           Suzuki  Grand Vitara  
2    Mercedes-Benz      SL-Class  
3             Ford        Escape  
4         Infiniti             G  
..             ...           ...  
563           MINI        Cooper  
564       Maserati  Quattroporte  
565     Mitsubishi        Lancer  
566            GMC   Sierra 3500  
567       Cadillac       DeVille  

[568 rows x 9 columns]

There are only 568 rows in the drivers_license table where the age is less than or equal to 21, compared to 9439 rows in the drivers_license table where the age is greater than 21.

In [20]:
# Select all records from table get_fit_now_check_in and order by check_in_date:

with engine.connect() as connection:   # Establish a connection
    query = text("""SELECT *
                    FROM get_fit_now_check_in
                    ORDER BY check_in_date
                    """) # Define the query
    practice9 = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database

# Print the results
practice9

membership_id  check_in_date  check_in_time  check_out_time
0            82GA2       20170101            431            1116
1            KE0FH       20170101            375             447
2            C84S4       20170101            580             864
3            XTE42       20170101           1188            1199
4            TQU5U       20170101            381            1069
...            ...            ...            ...             ...
2698         C581L       20180430             76             642
2699         0U51D       20180501             88             218
2700         7788A       20180501            532             705
2701         VHN8M       20180501            658             989
2702         MORMA       20180501           1029            1183

[2703 rows x 4 columns]

In [21]:
# Select all records from table get_fit_now_check_in and order by check_in_date and order by check_in_time:
# Do you see a difference in the order of the rows?

with engine.connect() as connection:   # Establish a connection
    query = text("""SELECT *
                    FROM get_fit_now_check_in
                    ORDER BY check_in_date, check_in_time
                    """) # Define the query
    practice10 = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database

# Print the results
practice10

membership_id  check_in_date  check_in_time  check_out_time
0            C581L       20170101             23            1071
1            3JP0X       20170101            154             609
2            J2033       20170101            242             955
3            KE0FH       20170101            375             447
4            TQU5U       20170101            381            1069
...            ...            ...            ...             ...
2698         133ED       20180430           1197            1200
2699         0U51D       20180501             88             218
2700         7788A       20180501            532             705
2701         VHN8M       20180501            658             989
2702         MORMA       20180501           1029            1183

[2703 rows x 4 columns]

The second table is now also sorted by check_in_time, so I do see a difference in the order of rows.

Now that we have practiced, let's solve this murder!

All we know at the start of our investigation is the <b>murder</b> occurred on <b>Jan.15, 2018</b> and that it took place in <b>SQL City<b>.

In [25]:
# Write a query in SQL to view the crime scene report 
# for the murder occurred on Jan.15, 2018 and that it took place in SQL City.
# Select ALL records FROM the table WHERE...

# Select all records from the table crime_scene_report where the date is 'Jan 15, 2018', the type is 'murder', and the city is 'SQL City'

with engine.connect() as connection:   # Establish a connection
    query = text("""SELECT *
                    FROM crime_scene_report
                    WHERE date='20180115' AND type='murder' AND city='SQL City'
                    """) # Define the query
    start = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database

# Print the results
print(start.to_string())

       date    type                                                                                                                                                                                description      city
0  20180115  murder  Security footage shows that there were 2 witnesses. The first witness lives at the last house on "Northwestern Dr". The second witness, named Annabel, lives somewhere on "Franklin Ave".  SQL City


Another clue! It looks like we have two witnesses! Now who are they and how do we find them?

In [27]:
# Write two queries in SQL to find each witness in the person table. 
# For the first query as first_witness, select all the records from person where the address_street_name is Northwestern Drive and order 
# by address_number
# Fetch the first_witness

# Select all records from the table person where address_street_name is 'Northwestern Dr' and then order by address_number

with engine.connect() as connection:   # Establish a connection
    query = text("""SELECT *
                    FROM person
                    WHERE address_street_name='Northwestern Dr'
                    ORDER BY address_number
                    """) # Define the query
    first_witness = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database

# Print the results
first_witness

id                  name  license_id  address_number  \
0   89906       Kinsey Erickson      510019             309   
1   15171          Weldon Penso      336999             311   
2   40336       Fredia Billeter      251082             404   
3   85280             Donn Duin      569122             540   
4   12711        Norman Apolito      667757             599   
5   75484          Del Tacderen      447887             607   
6   26758      Angelena Billman      942074             629   
7   39688            Abe Roeker      537445             689   
8   10010             Muoi Cary      385336             741   
9   40760          Renata Faver      976650             745   
10  25615         Olevia Morena      388477             812   
11  99826          Ivy Kazarian      994470             892   
12  44542        Ivory Mcgrevey      873842             897   
13  55130        Shiloh Huntoon      472118             951   
14  33415         Homer Simpson      108392             981   
15  29608       Willetta Lurvey      800568            1028   
16  18376              Josh Shi      653712            1091   
17  22239         Dusty Sigafus      710517            1125   
18  70485     Andera Kostelecky      354822            1252   
19  23960  Kristopher Lagerberg      658777            1392   
20  71924        Laurie Scronce      774116            1418   
21  81052     Calvin Czelusniak      471483            1516   
22  72076     Courtney Bordeaux      324614            1707   
23  51552           Bonita Knie      764124            1726   
24  88423           Teri Ehrich      510876            1879   
25  81885         Kurtis Masuda      934201            2108   
26  96178           Galina Khuu      450248            2183   
27  80921       Donovan Yackley      222395            2386   
28  34352    Vincenza Burkhardt      436470            2512   
29  53699      Christian Pagano      628433            2585   
30  54564         Sammie Sengun      217286            2590   
31  90397     Dionna Kranwinkle      797260            2678   
32  34155      Allyson Lazenson      814209            2840   
33  59307          Cyril Yongue      753524            2903   
34  98593           Jonah Toner      578982            2947   
35  44004           Alison Eska      327797            2951   
36  68690            Yer Modest      852404            3046   
37  51114     Christena Saffell      814284            3055   
38  23044          Val Portlock      924989            3143   
39  28360        Rashad Cascone      924517            3212   
40  53076           Boris Bijou      664914            3327   
41  36378        Freddie Ellzey      267882            3449   
42  87456   Leonora Wolfsberger      215868            3483   
43  93509        Emmitt Aceuedo      916706            3491   
44  19420           Cody Schiel      890431            3524   
45  96595         Coretta Cubie      303645            3631   
46  73368        Torie Thalmann      773862            3697   
47  53890        Sophie Tiberio      957671            3755   
48  17729        Lasonya Wildey      439686            3824   
49  14887        Morty Schapiro      118009            4919   

   address_street_name        ssn  
0      Northwestern Dr  635287661  
1      Northwestern Dr  131379495  
2      Northwestern Dr  108056488  
3      Northwestern Dr  613873810  
4      Northwestern Dr  778264744  
5      Northwestern Dr  243821337  
6      Northwestern Dr  341854756  
7      Northwestern Dr  438969447  
8      Northwestern Dr  828638512  
9      Northwestern Dr  547572040  
10     Northwestern Dr  962831331  
11     Northwestern Dr  208359813  
12     Northwestern Dr  988557407  
13     Northwestern Dr  826647644  
14     Northwestern Dr  871392019  
15     Northwestern Dr  843751378  
16     Northwestern Dr  193899001  
17     Northwestern Dr  724386723  
18     Northwestern Dr  787423012  
19     Northwestern Dr  492912529  
20     Northwestern Dr  254883636  
2

We have identified our first witness as ___Morty Schapiro___. Now lets find the second!

In [29]:
# For the third query as second_witness,
# Select all the records from person where the address_street_name is Franklin Ave and the name includes Annabel. 
# Fetch the second_witness

# Select all records from the table person where address_street_name is 'Franklin Ave' and the name contains 'Annabel'

with engine.connect() as connection:   # Establish a connection
    query = text("""SELECT *
                    FROM person
                    WHERE address_street_name='Franklin Ave' AND name LIKE '%Annabel%'
                    """) # Define the query
    second_witness = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database

# Print the results
second_witness

id            name  license_id  address_number address_street_name  \
0  16371  Annabel Miller      490173             103        Franklin Ave   

         ssn  
0  318771143

We have identified our second witness as ___Annabel Miller___. What were their witness statements? What did they see?

In [31]:
# For our fourth query, witness_statements, lets find out what each witness saw...or claim they saw?
# Select the person's name and interview where the person_id is X, X (Get the ID from the above queries)

# Select all records from the table interview where the person_id is '14887' or '16371'

with engine.connect() as connection:   # Establish a connection
    query = text("""SELECT *
                    FROM interview
                    WHERE person_id IN ('14887', '16371')
                    """)
    witness_statements = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database

# Print the results
print(witness_statements.to_string())

   person_id                                                                                                                                                                                                                       transcript
0      14887  I heard a gunshot and then saw a man run out. He had a "Get Fit Now Gym" bag. The membership number on the bag started with "48Z". Only gold members have those bags. The man got into a car with a plate that included "H42W".
1      16371                                                                                                            I saw the murder happen, and I recognized the killer from my gym when I was working out last week on January the 9th.


A few new clues! Explore the different clues with multiple queries. Seems like a reliable statement!

In [33]:
# Create a query to select for the following clues: (You can create multiple queries or combine them into one and go from there.)
# Remember to write our your query in the comments!

# Select all records from the table get_fit_now_check_in where the membership_id starts with '48Z' and the check_in_date is 'Jan 9, 2018'

with engine.connect() as connection:   # Establish a connection
    query = text("""SELECT *
                    FROM get_fit_now_check_in
                    WHERE membership_id LIKE '48Z%' AND check_in_date='20180109'
                    """) # Define the query
    suspect = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database

# Print the results
suspect

membership_id  check_in_date  check_in_time  check_out_time
0         48Z7A       20180109           1600            1730
1         48Z55       20180109           1530            1700

In [34]:
# Select all records from the table drivers_license where the plate_number includes 'H42W' and the gender is 'male'

with engine.connect() as connection:   # Establish a connection
    query = text("""SELECT *
                    FROM drivers_license
                    WHERE plate_number LIKE '%H42W%' AND gender='male'
                    """) # Define the query
    suspect2 = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database

# Print the results
suspect2

id  age  height eye_color hair_color gender plate_number   car_make  \
0  423327   30      70     brown      brown   male       0H42W2  Chevrolet   
1  664760   21      71     black      black   male       4H42WR     Nissan   

  car_model  
0  Spark LS  
1    Altima

In [35]:
# Select all records from the table person where the license_id is '423327' or '664760'

with engine.connect() as connection:   # Establish a connection
    query = text("""SELECT *
                    FROM person
                    WHERE license_id IN ('423327', '664760')
                    """) # Define the query
    suspect3 = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database

# Print the results
suspect3

id            name  license_id  address_number    address_street_name  \
0  51739  Tushar Chandra      664760             312                 Phi St   
1  67318   Jeremy Bowers      423327             530  Washington Pl, Apt 3A   

         ssn  
0  137882671  
1  871539279

In [36]:
# Select all records from the table get_fit_now_member where the person_id is '51739' or '67318'

with engine.connect() as connection:   # Establish a connection
    query = text("""SELECT *
                    FROM get_fit_now_member
                    WHERE person_id IN ('51739', '67318')
                    """) # Define the query
    suspect4 = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database

# Print the results
suspect4

id  person_id           name  membership_start_date membership_status
0  48Z55      67318  Jeremy Bowers               20160101              gold

We have a suspect! Lets check their witness statement. Do they have an alibi?

In [38]:
# Remember to write your query in the comments!

# Select all records from the table interview where the person_id is '67318'

with engine.connect() as connection:   # Establish a connection
    query = text("""SELECT *
                    FROM interview
                    WHERE person_id='67318'
                    """) # Define the query
    alibi = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database

# Print the results
print(alibi.to_string())

   person_id                                                                                                                                                                                                                                          transcript
0      67318  I was hired by a woman with a lot of money. I don't know her name but I know she's around 5'5" (65") or 5'7" (67"). She has red hair and she drives a Tesla Model S. I know that she attended the SQL Symphony Concert 3 times in December 2017.\n


Uh oh...turns out it wasn't just one person behind the crime...who is this additional person involved?

In [40]:
# Remember to write our your query in the comments!
# You can complete this in a single query or multiple queries.

# Select all records from the table drivers_license where 
# the height is between 65 and 67, the hair_color is 'red' and the car_model is 'Model_S'

with engine.connect() as connection:   # Establish a connection
    query = text("""SELECT *
                    FROM drivers_license
                    WHERE height BETWEEN 65 AND 67 
                    AND hair_color='red' AND car_model='Model S'
                    """) # Define the query
    accomplice = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database

# Print the results
accomplice

id  age  height eye_color hair_color  gender plate_number car_make  \
0  202298   68      66     green        red  female       500123    Tesla   
1  291182   65      66      blue        red  female       08CM64    Tesla   
2  918773   48      65     black        red  female       917UU3    Tesla   

  car_model  
0   Model S  
1   Model S  
2   Model S

Three different people match this description! Now which one attended the SQL Symphony Concert three times? 

In [42]:
# Let's find the killer!
# Remember to write our your query in the comments!

# Select all records from the table person where the license_id is '202298', '291182', or '918773'

with engine.connect() as connection:   # Establish a connection
    query = text("""SELECT *
                    FROM person
                    WHERE license_id IN ('202298', '291182', '918773')
                    """) # Define the query
    accomplice2 = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database

# Print the results
accomplice2

id              name  license_id  address_number address_street_name  \
0  78881          Red Korb      918773             107         Camerata Dr   
1  90700     Regina George      291182             332           Maple Ave   
2  99716  Miranda Priestly      202298            1883          Golden Ave   

         ssn  
0  961388910  
1  337169072  
2  987756388

In [43]:
# Select all records from the table facebook_event_checkin where the person_id is '78881', '90700', or '99716'

with engine.connect() as connection:   # Establish a connection
    query = text("""SELECT *
                    FROM facebook_event_checkin
                    WHERE person_id IN ('78881', '90700', '99716')
                    """) # Define the query
    accomplice3 = pd.read_sql(query, connection) # Use pandas to read the sql query with the connection to the database

# Print the results
accomplice3

person_id  event_id            event_name      date
0      99716      1143  SQL Symphony Concert  20171206
1      99716      1143  SQL Symphony Concert  20171212
2      99716      1143  SQL Symphony Concert  20171229

<b>The murderer(s) is/are:</b> Jeremy Bowers and Miranda Priestly

In [45]:
# Disconnect from the database. Always remember to disconnect :) 
connection.close()